test/test_loopy.py

```diff
@@ -2886,6 +2886,39 @@ def test_empty_domain(ctx_factory, tag):
     assert (c.get() == 0).all()
 
 
+def test_if_empty_domain(ctx_factory):
+    ctx = ctx_factory()
+    queue = cl.CommandQueue(ctx)
+
+    prg = lp.make_kernel(
+            "{[i, j, k]: 0 <= i < n and 0 <= j < m and 0 <= k < p}",
```

Comment:
This is a corner of the semantics that I'm repeatedly unsure about, but I've talked myself into believing that what it does is correct. You wrote a set of statement instances, and, mathematically speaking, if `p == 0`, that set is empty. If you'd like "imperfect loop nesting" then you need to do as @kaushikcfd suggests: split the domain.

I'm still curious to hear: how sold are you on this argument?

cc @jdsteve2

Reply:
I definitely agree this is a degenerate case. One thing I find a bit unintuitive, is that the limit `p -> 0` gives very different behavior (i.e. nothing gets filled in). Separating the `if` for `p` and `m` there so that each one is around its own loop seems a bit more expected, while working the same for the case `p, m > 0`.

Maybe just a way to warn / error if a dimension is collapsed?

Reply:
Semantically/mathematically, our hands are tied here:
``python
>>> import islpy as isl
>>> s = isl.Set("[m,n,p] -> {[i, j, k]: 0 <= i < n and 0 <= j < m and 0 <= k < p}")
>>> s & isl.Set("[m,n,p] -> {[i,j,k]: p = 0}")
Set("[m, n, p] -> { [i, j, k] : false }")
```
As soon as `p=0`, that set (of statement instances) is the empty set, end of story.

Do you mean we should warn about any loop domain definition defining more than one iname? That makes loop interchange awfully hard. Plus, OpenCL "grids" work exactly this way: if any dimension is zero-length, the set is empty.

Reply:
I just meant that it would have been nice to have some sort of warning going "that 'if' is always false. are you sure you want to do this?". Although that sounds like something that's only known once all the sizes are given.

As for the rest, I'm starting to think it's just my lack of familiarity with thinking of more complex loops as sets. For example, I would expect the OpenCL "empty grid" look-alike to be
```
{[i, j]: 0<=i<n and 0<=j<p}
result[i,j] = 1
```
and it should do nothing if `p = 0`. However, having
```
{[i,j,k]: 0<=i<n and 0<=j<m and m<=k<p}
result[i,j]=1
result[i,k]=2
```
do nothing for `m > 0` and `p = m` is a bit weird.

Reply:
Just to be clear, I agree that the current behavior is ok given the fairly clear cut math semantics. 

My only complaint is that I wished something would have yelled at me that the kernel is a noop when `p = 0`. I'm happy to close this, since that's also my fault for not checking everything was initialized (in the real version of that code).

Reply:
Thanks for thinking about this. I think the warning would be difficult to pull off in the general case. (Difficult enough, at least at high performance, that I can't think of a good way of doing it.) I'll close this, but if I think of a good way to address it, I might reopen this later.